cirq/protocols/apply_mixture_protocol.py

```diff
@@ -0,0 +1,309 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A protocol for implementing high performance mixture evolutions."""
+
+from typing import Any, Iterable, Optional, TypeVar, Union
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq import linalg
+from cirq.protocols.apply_unitary_protocol import (
+    apply_unitary,
+    ApplyUnitaryArgs,
+)
+
+from cirq.protocols.mixture import mixture
+from cirq.protocols import qid_shape_protocol
+from cirq.type_workarounds import NotImplementedType
+
+# This is a special indicator value used by the apply_mixture method
+# to determine whether or not the caller provided a 'default' argument. It must
+# be of type np.ndarray to ensure the method has the correct type signature in
+# that case. It is checked for using `is`, so it won't have a false positive if
+# the user provides a different np.array([]) value.
+
+RaiseTypeErrorIfNotProvided = np.array([])  # type: np.ndarray
+
+TDefault = TypeVar('TDefault')
+
+
+class ApplyMixtureArgs:
+    """Arguments for performing a mixture of unitaries.
+
+    The receiving object is expected to mutate `target_tensor` so that it
+    contains the density matrix after applying the mixture then return
+    `target_tensor`. Alternatively, if workspace is required,
+    the receiving object can overwrite `out_buffer` with the results
+    and return `out_buffer`. Or, if the receiving object is attempting to
+    be simple instead of fast, it can create an entirely new array and
+    return that.
+
+    Attributes:
+        target_tensor: The input tensor that needs to be left and right
+            multiplied and summed, representing the effect of the mixture.
+            The tensor will have the shape (2, 2, 2, ..., 2). It usually
+            corresponds to a multi-qubit density matrix, with the first
+            n indices corresponding to the rows of the density matrix and
+            the last n indices corresponding to the columns of the density
+            matrix.
+        out_buffer: Pre-allocated workspace with the same shape and
+            dtype as the target tensor. If buffers are used, the result should
+            end up in this buffer. It is the responsibility of calling code
+            to notice if the result is this buffer.
+        auxiliary_buffer0: Pre-allocated workspace with the same shape and dtype
+            as the target tensor.
+        auxiliary_buffer1: Pre-allocated workspace with the same shape
+            and dtype as the target tensor.
+        left_axes: Which axes to multiply the left action of the mixture upon.
+        right_axes: Which axes to multiply the right action of the mixture upon.
+    """
+
+    def __init__(self, target_tensor: np.ndarray, out_buffer: np.ndarray,
+                 auxiliary_buffer0: np.ndarray, auxiliary_buffer1: np.ndarray,
+                 left_axes: Iterable[int], right_axes: Iterable[int]):
```

Comment:
We should make `right_axes` optional, in order to support the use case of applying to a state vector. For consistency we could also make `left_axes` optional.

Reply:
fixed